### PR TITLE
Use interpolated strings in place of `StringBuilder.AppendFormat` where possible.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
This PR converts all calls to `StringBuilder.AppendFormat` in non-test code to instead use `StringBuilder.Append` with interpolated strings, taking advantage of the .NET 6 interpolated string handlers, that allocate zero intermediate objects, writing the characters directly to the `StringBuilder` in our case.

Call sites where the format strings were obtained from resources could not be optimized and were not changed.

Earlier frameworks might see a slightly decreased performance but the changed code does not seem to be performance-critical.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/CallSiteJsonFormatter.cs [hunk 1]
```diff
@@ -159,7 +159,7 @@ public void StartProperty(string name)
                 {
                     _firstItem = false;
                 }
-                Builder.AppendFormat("\"{0}\":", name);
+                Builder.Append($"\"{name}\":");
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/CallSiteJsonFormatter.cs:162]: @eerhardt, opinions on taking changes like this that will improve netcoreappcurrent but add cost to downlevel?

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs [hunk 2]
```diff
@@ -414,7 +414,7 @@ private string FormatTags(KeyValuePair<string, string>[] labels)
                 StringBuilder sb = new StringBuilder();
                 for (int i = 0; i < labels.Length; i++)
                 {
-                    sb.AppendFormat(CultureInfo.InvariantCulture, "{0}={1}", labels[i].Key, labels[i].Value);
+                    sb.Append(CultureInfo.InvariantCulture, $"{labels[i].Key}={labels[i].Value}");
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:417]: This library compiles for more than just netcoreappcurrent, and this overload doesn't exist earlier than .NET 6.  Please revert all of the changes to this library.

File: src/libraries/System.Speech/src/Internal/Synthesis/TextFragmentEngine.cs [hunk 3]
```diff
@@ -195,10 +195,10 @@ public void ProcessUnknownElement(object voice, ref FragmentState fragmentState,
         public void StartProcessUnknownAttributes(object voice, ref FragmentState fragmentState, string element, List<SsmlXmlAttribute> extraAttributes)
         {
             StringBuilder sb = new();
-            sb.AppendFormat(CultureInfo.InvariantCulture, "<{0}", element);
```
Comment (stephentoub) [inline src/libraries/System.Speech/src/Internal/Synthesis/TextFragmentEngine.cs:198]: Please revert all of the changes to System.Speech.  We've been avoiding churning the code, as it's here only for compat and doesn't have a robust set of tests.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs (hunk 2 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:417]: How about I add an `#if NET_6_0_OR_GREATER`?

Comment (stephentoub) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:417]: Can you share the benchmark that indicates it's worth the added complexity and maintenance burden?

Comment (teo-tsirpanis) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:417]: OK, never mind, I will revert it.

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/CallSiteJsonFormatter.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/CallSiteJsonFormatter.cs:162]: > Thinking about our customers though, what is our recommendation for a "normal multi-targeted library"?

The same for any new APIs: decide whether the extra maintenance cost of maintaining two different code paths is worth the benefits.  In the case of new APIs, the alternate being weighed is not getting the benefit at all up-level.  In this case, the alternate being weighed is paying more downlevel.

File: src/libraries/System.Speech/src/Internal/Synthesis/TextFragmentEngine.cs (hunk 3 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Speech/src/Internal/Synthesis/TextFragmentEngine.cs]: Done.

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/CallSiteJsonFormatter.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/CallSiteJsonFormatter.cs:162]: All 3 places being changed in DependencyInjection are in non-performance-critical paths. 2 in EventSource logging, 1 in building an exception message.

I guess just like the other libraries, I would revert these changes unless we had data backing up why we should make them.

Comment (teo-tsirpanis) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/CallSiteJsonFormatter.cs:162]: Oh, OK then, I will revert them in DI as well.

**Review by stephentoub — APPROVED** [review]
Thanks.

**Review by deeprobin — APPROVED** [review]
Good job!

**Pull request merged**